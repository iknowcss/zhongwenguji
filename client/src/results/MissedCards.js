import React from 'react';
import style from './MissedCards.module.scss';
import connect from 'react-redux/es/connect/connect';
import mapSelectors from '../util/mapSelectors';
import { missedCards } from '../characterTest/characterTestReducer';
import {loadSamples} from '../characterTest/characterTestActions';
import formatPinyin from '../util/formatPinyin';
import Button from '../component/Button';

function pairPinyinDefinitions(pinyinArray, definitionArray) {
  const result = [];
  for (let i = 0; i < pinyinArray.length; i++) {
    const pinyin = pinyinArray[i];
    const definition = [definitionArray[i]];
    result.push({ pinyin, definition })
  }
  return result;
}

function formatDefinition(definition) {
  return definition.replace(/\//g, ' / ');
}

const MissedCards = ({ missedCards = [] } = {}) => {
  return (
    <div className={style.container}>
      <div className={style.cardRowContainer}>
      {missedCards.map((card) => (
        <div key={card.index} className={style.cardRowColor}>
          <div className={style.cardRow}>
            <div className={style.cardCharacterColumn}>
              <span>{card.character}</span>
            </div>
            <div className={style.cardPinyinDefColumn}>
              {pairPinyinDefinitions(card.pinyin, card.definition).map(({ pinyin, definition }, i) => (
                <div key={i} className={style.cardPinyinDefRow}>
                  <div className={style.cardPinyinColumn}>
                    <span>{formatPinyin(pinyin)}</span>
                  </div>
                  <div className={style.cardDefColumn}>
                    <span>{formatDefinition(definition[0])}</span>
                  </div>
                </div>
              ))}
            </div>
          </div>
        </div>
      ))}
      </div>
      <div className={style.actionButtonContainer}>
        <Button secondary className={style.actionButton}>Back</Button>
        <Button className={style.actionButton}>Add to Skritter</Button>
      </div>
    </div>
  );
};

export { MissedCards as Pure };

export default connect(
  mapSelectors({ missedCards }),
  { loadSamples }
)(MissedCards);
