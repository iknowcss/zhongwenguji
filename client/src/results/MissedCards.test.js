import React from 'react';
import Renderer from 'react-test-renderer';
import { Pure as MissedCards } from './MissedCards';

describe('MissedCards', () => {
  let component;

  function setup(props) {
    component = Renderer.create(<MissedCards {...props} />);
  }

  it('renders no cards', () => {
    setup();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders cards', () => {
    setup({
      missedCards: [
        {
          index: 1,
          character: 'è¯»',
          pinyin: ['dou4', 'du2'],
          definition: [
            'comma/phrase marked by pause',
            'to read/to study'
          ]
        }
      ]
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('returns to the results page', () => {
    const preventDefault = jest.fn();
    const showTestResults = jest.fn();
    setup({ showTestResults });
    const button = component.root.findByProps({ stringId: 'reviewMissed.backButton' });
    button.props.onClick({ preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(showTestResults).toHaveBeenCalledTimes(1);
  });
});
