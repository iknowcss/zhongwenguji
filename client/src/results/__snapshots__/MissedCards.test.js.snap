// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissedCards renders cards 1`] = `
<div
  className="container"
>
  <div
    className="actionButtonContainer"
  >
    <button
      className="button actionButton"
      onClick={[Function]}
    >
      reviewMissed.backButton
    </button>
    <button
      className="button actionButton"
      onClick={[Function]}
    >
      Add to Skritter
    </button>
  </div>
  <div
    className="cardRowContainer"
  >
    <div
      className="cardRowColor"
    >
      <div
        className="cardRow"
      >
        <div
          className="cardCharacterColumn"
        >
          <span>
            读
          </span>
        </div>
        <div
          className="cardPinyinDefColumn"
        >
          <div
            className="cardPinyinDefRow"
          >
            <div
              className="cardPinyinColumn"
            >
              <span>
                dòu
              </span>
            </div>
            <div
              className="cardDefColumn"
            >
              <span>
                comma / phrase marked by pause
              </span>
            </div>
          </div>
          <div
            className="cardPinyinDefRow"
          >
            <div
              className="cardPinyinColumn"
            >
              <span>
                dú
              </span>
            </div>
            <div
              className="cardDefColumn"
            >
              <span>
                to read / to study
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MissedCards renders no cards 1`] = `
<div
  className="container"
>
  <div
    className="actionButtonContainer"
  >
    <button
      className="button actionButton"
      onClick={[Function]}
    >
      reviewMissed.backButton
    </button>
    <button
      className="button actionButton"
      onClick={[Function]}
    >
      Add to Skritter
    </button>
  </div>
  <div
    className="cardRowContainer"
  />
</div>
`;
